// @xdn/cli/commands/run.js
"use strict";

const detectFramework = require('../frameworks/detectFramework');

const run = require('../utils/run');

const {
  join
} = require('path');

const {
  existsSync
} = require('fs');

const chalk = require('chalk');

const get = require('lodash/get');

const resolveInPackage = require('../utils/resolveInPackage');

const logo = require('../utils/logo');

exports.command = 'run';
exports.describe = 'Runs your project locally, simulating the XDN cloud environment';
exports.builder = {
  production: {
    type: 'boolean',
    alias: 'p',
    describe: 'Runs a production build of your app.'
  },
  serverless: {
    type: 'boolean',
    alias: 's',
    describe: 'Runs a production build of your app using serverless-offline. This is the best way to simulate the behavior of running your app in the cloud.'
  },
  cache: {
    type: 'boolean',
    alias: 'c',
    describe: 'Enables caching in development mode. Caching is enabled by default in production mode.'
  }
};

exports.handler = async ({
  context,
  production,
  serverless,
  cache
}) => {
  const framework = await detectFramework();
  const xdnDir = join(process.cwd(), '.xdn');
  const {
    logger
  } = context;

  if (process.env.NODE_ENV === 'production' || production || serverless) {
    if (!existsSync(xdnDir)) {
      logger.error(`\nXDN production build not found. Please run ${chalk.green('xdn build')} before running ${chalk.green('xdn run --production')}.\n`);
      process.exit(1);
    }

    process.env.NODE_ENV = 'production';
    process.env.XDN_CACHE = 'true';
    production = true;
  }

  if (serverless) {
    logger.info(`> Running ${logo} in serverless mode using a production build...`);
    await Promise.all([run(resolveInPackage('serverless', 'bin', 'serverless'), ['offline'], {
      cwd: join(xdnDir, 'lambda')
    }), run('node', [resolveInPackage('@xdn/core', 'bin', 'serveAssets.js')])]);
  } else if (production) {
    logger.info(`> Running ${logo} in production mode...`);
    await run(getNodeCommand(), [resolveInPackage(get(framework, 'builder', '@xdn/core'), 'bin', 'run.js')]);
  } else {
    logger.info(`> Running ${logo} in development mode with caching ${cache ? chalk.green('enabled') : chalk.red('disabled')}`);

    if (cache) {
      process.env.XDN_CACHE = 'true';
    }

    await run(getNodeCommand(), [resolveInPackage(get(framework, 'builder', '@xdn/core'), 'bin', 'run.js')]);
  }
};
/**
 * Returns babel-node if installed, otherwise node
 */


function getNodeCommand() {
  try {
    return resolveInPackage('@babel/node', 'bin', 'babel-node.js');
  } catch (e) {
    return 'node';
  }
}
