// @xdn/cli/commands/use.js
"use strict";

const XdnPackageJson = require('../utils/XdnPackageJson');

const logo = require('../utils/logo');

const {
  installDependencies
} = require('../utils/packageManager');

exports.command = 'use [xdnVersion]';
exports.describe = 'Updates XDN version';
exports.builder = {
  xdnVersion: {
    type: 'string',
    describe: 'Updates XDN version',
    demandOption: true
  },
  path: {
    type: 'string',
    describe: "Path to your site's root director. Uses current directory by default",
    default: '.'
  }
};

exports.handler = async ({
  context,
  xdnVersion,
  path: projectPath
}) => {
  const packageJson = new XdnPackageJson(projectPath);
  await context.logger.step(`Updating ${logo}`, async () => {
    context.logger.info(`> Current: ${packageJson.findCurrentXdnVersion()}`);
    context.logger.info(`> New: ${xdnVersion}`);
    const devDependencies = Object.fromEntries(packageJson.xdnDevDependencies().map(([name]) => [name, xdnVersion]));
    const runtimeDependencies = Object.fromEntries(packageJson.xdnRuntimeDependencies().map(([name]) => [name, xdnVersion]));
    await installDependencies(runtimeDependencies, {
      dev: false,
      overrideInstalled: true
    });
    await installDependencies(devDependencies, {
      dev: true,
      overrideInstalled: true
    });
  });
};
