// @xdn/core/lambda/JSBackend.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var tcp_port_used_1 = __importDefault(require("tcp-port-used"));
var started = false;
/**
 * Base class that all framework integration packages should implement to start a
 * server instance of the framework to which requests can be proxied as the __js__ backend.
 */
var JSBackend = /** @class */ (function () {
    /**
     * @param port The port to which the server instance should bind.
     */
    function JSBackend(port) {
        this.started = false;
        this.port = port;
    }
    /**
     * Returns true if the provided class extends JSBackend.
     * @param obj A class
     */
    JSBackend.isSubclass = function (obj) {
        return obj.__type__ == 'JSBackend';
    };
    /**
     * Starts the app's server build. We don't start the app if the target port is already
     * taken. This will be the case when using xdn run --serverless to emulate the cloud locally,
     * since serverless-offline sandboxes the app in such a way that it's difficult to keep
     * track of whether or not the app has been started - checking the port is the easiest way -
     * and it will attempt to start the app on every request.
     */
    JSBackend.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = started;
                        if (_a) return [3 /*break*/, 2];
                        return [4 /*yield*/, tcp_port_used_1.default.check(this.port)];
                    case 1:
                        _a = (_b.sent());
                        _b.label = 2;
                    case 2:
                        if (!_a) return [3 /*break*/, 3];
                        return [2 /*return*/];
                    case 3:
                        started = true;
                        return [4 /*yield*/, this.doStart()];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Subclasses must override this method to implement starting the app's server build.
     */
    JSBackend.prototype.doStart = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('Subclasses must override the doStart method.');
            });
        });
    };
    JSBackend.__type__ = 'JSBackend';
    return JSBackend;
}());
exports.default = JSBackend;
