// @xdn/cli/utils/errors.js
"use strict";

/**
 * XdnCLIErrors are log in different ways:
 * - 'userMessage' is displayed in red bold in CLI output and added to LD logs
 * - 'errorDetails' is displayed in red normal in CLI output and added to LD logs
 * - 'addToLeDeployerLogs' is only added to LD logs.
 *
 */
class XdnCLIError extends Error {
  constructor(userMessage, {
    errorDetails,
    addToLeDeployerLogs
  } = {}) {
    super(userMessage);
    this.errorDetails = errorDetails;
    this.addToLeDeployerLogs = addToLeDeployerLogs;
  }

}

class XdnCLIUserError extends XdnCLIError {}

class BadRequestError extends XdnCLIUserError {}

class AuthenticationError extends XdnCLIUserError {}

class MissingPackageJsonError extends XdnCLIUserError {}

class MissingXDNPackagesError extends XdnCLIUserError {}

class DeployError extends XdnCLIUserError {}

class ValidationError extends XdnCLIUserError {}

class LeDeployerApiError extends XdnCLIUserError {}

exports.XdnCLIError = XdnCLIError;
exports.BadRequestError = BadRequestError;
exports.AuthenticationError = AuthenticationError;
exports.MissingPackageJsonError = MissingPackageJsonError;
exports.MissingXDNPackagesError = MissingXDNPackagesError;
exports.DeployError = DeployError;
exports.LeDeployerApiError = LeDeployerApiError;
exports.ValidationError = ValidationError;
