// @xdn/cli/utils/XdnPackageJson.js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const path = require('path');

const {
  MissingPackageJsonError,
  MissingXDNPackagesError
} = require('./errors');

const isEmpty = require('lodash/isEmpty');

const getValue = require('lodash/get');

class XdnPackageJson {
  /**
   * Creates class instance for packageJson manipulation
   * @param {String} givenPath path to project working directory
   */
  constructor(givenPath = '.') {
    _defineProperty(this, "XDN_PACKAGE_PREFIX", '@xdn/');

    _defineProperty(this, "package", void 0);

    _defineProperty(this, "path", void 0);

    // Absoulte path to packageJson
    const absolutePath = path.resolve(process.cwd(), givenPath);
    this.path = path.join(absolutePath, 'package.json'); // Load Package.json

    this.load();
  }
  /**
   * Loads package.json
   */


  load() {
    try {
      this.package = require(this.path);
      return this.package;
    } catch (e) {
      throw new MissingPackageJsonError(`Failed to locate package.json in ${this.path}`);
    }
  }
  /**
   *  Returns a compiled list of dev and runtime xdn packages
   */


  xdnAllPackages() {
    return [...this.xdnRuntimeDependencies(), ...this.xdnDevDependencies()];
  }
  /**
   * Returns a xdn version found in package.json
   */


  findCurrentXdnVersion() {
    const allPackages = this.xdnAllPackages();

    if (isEmpty(allPackages)) {
      throw new MissingXDNPackagesError(`There is no xdn packages installed in ${this.path}`);
    }

    return allPackages[0][1];
  }
  /**
   * Return list of xdn runtime dependencies
   */


  xdnRuntimeDependencies() {
    return Object.entries(this.find('dependencies', {})).filter(([name]) => name.startsWith(this.XDN_PACKAGE_PREFIX));
  }
  /**
   * Returns list of xdn dev dependencies
   */


  xdnDevDependencies() {
    return Object.entries(this.find('devDependencies', {})).filter(([name]) => name.startsWith(this.XDN_PACKAGE_PREFIX));
  }
  /**
   * Returns package json value by path
   * @param {String} path Path to object key (loash.get)
   */


  find(path, defaultValue = undefined) {
    return getValue(this.package, path, defaultValue);
  }
  /**
   * Loads package.json
   * @param {String} givenPath
   */


  static loadPackageJson(givenPath) {
    return new XdnPackageJson(givenPath).package;
  }

}

module.exports = XdnPackageJson;
