// @xdn/prefetch/sw/Prefetcher.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var workbox_routing_1 = require("workbox-routing");
var workbox_strategies_1 = require("workbox-strategies");
var workbox_expiration_1 = require("workbox-expiration");
var workbox_cacheable_response_1 = require("workbox-cacheable-response");
var constants_1 = require("../constants");
var log_1 = __importDefault(require("./log"));
var PrefetchPlugin_1 = __importDefault(require("./PrefetchPlugin"));
var messageBrowser_1 = __importDefault(require("./messageBrowser"));
var ChangeMethodPlugin_1 = __importDefault(require("./ChangeMethodPlugin"));
var Stats = /** @class */ (function () {
    function Stats() {
        this.hits = 0;
        this.misses = 0;
    }
    return Stats;
}());
exports.Stats = Stats;
/**
 * Handles prefetching and caching in the the service worker.  Responses
 * are cached based on the `browser.serviceWorkerSeconds` cache setting in your routes.
 *
 * **Example routes.js**
 *
 * ```js
 *  import { Router } from '@xdn/core/router'
 *
 *  module.exports = new Router()
 *    .match('/some/path', ({ cache }) => {
 *      cache({
 *        edge: {
 *          maxAgeSeconds: 60 * 60 * 24 // one day
 *        },
 *        browser: {
 *          serviceWorkerSeconds: 60 * 60 // one hour
 *        }
 *      })
 *    })
 * ```
 *
 * **Example Service Worker**
 *
 * ```js
 *  import { skipWaiting, clientsClaim } from 'workbox-core'
 *  import { Prefetcher } from '@xdn/prefetch/sw'
 *  import { precacheAndRoute } from 'workbox-precaching'
 *
 *  skipWaiting()
 *  clientsClaim()
 *  precacheAndRoute(self.__WB_MANIFEST || [])
 *
 *  new Prefetcher().route()
 * ```
 */
var Prefetcher = /** @class */ (function () {
    function Prefetcher(config) {
        if (config === void 0) { config = { plugins: [] }; }
        this.config = config;
        var manifestURL = '__xdn__/cache-manifest.js';
        if (config.cacheHost) {
            this.cacheHost = new URL("https://" + config.cacheHost);
            manifestURL = this.cacheHost.toString() + manifestURL;
        }
        importScripts(manifestURL);
        this.cacheableResponsePlugin = new workbox_cacheable_response_1.CacheableResponsePlugin({
            statuses: [200, 301, 302],
        });
        this.stats = new Stats();
        this.initStats();
    }
    /**
     * Adds a fetch event listener that serves content from the cache when available.  When
     * a request is not in the cache, it will be fetched from the network and added to the
     * cache if the route has `browser.serviceWorkerSeconds` defined in its `cache` setting.
     */
    Prefetcher.prototype.route = function () {
        var e_1, _a;
        var _b, _c;
        // @ts-ignore
        var manifest = self.__XDN_CACHE_MANIFEST__;
        try {
            for (var manifest_1 = __values(manifest), manifest_1_1 = manifest_1.next(); !manifest_1_1.done; manifest_1_1 = manifest_1.next()) {
                var _d = manifest_1_1.value, route = _d.route, cacheOptions = _d.cacheOptions;
                var maxAgeSeconds = (_b = cacheOptions.browser) === null || _b === void 0 ? void 0 : _b.serviceWorkerSeconds;
                var method = (_c = cacheOptions.browser) === null || _c === void 0 ? void 0 : _c.asMethod;
                if (maxAgeSeconds > 0) {
                    log_1.default("adding route " + route + ", maxAgeSeconds: " + maxAgeSeconds);
                    this.createRoute(new RegExp(route, 'i'), {
                        method: method,
                        maxAgeSeconds: maxAgeSeconds,
                        isPartialPathRoute: true,
                    });
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (manifest_1_1 && !manifest_1_1.done && (_a = manifest_1.return)) _a.call(manifest_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return this;
    };
    Prefetcher.prototype.cache = function (pattern, maxAgeSeconds) {
        if (maxAgeSeconds === void 0) { maxAgeSeconds = 60 * 60 * 24; }
        this.createRoute(pattern, { maxAgeSeconds: maxAgeSeconds, isPartialPathRoute: false });
        return this;
    };
    /**
     * Starts listening for messages from the browser
     */
    Prefetcher.prototype.initStats = function () {
        var _this = this;
        self.addEventListener('message', function (e) {
            if (e.data.action === 'get-stats') {
                messageBrowser_1.default({
                    type: 'stats',
                    stats: _this.stats,
                });
            }
        });
    };
    /**
     * If there's a cacheHost configured check whether the pattern to test
     * against is a partial pathname pattern. If so, check that the origin is
     * the same as the cacheHost
     * @param pattern A route pattern
     * @param isPartialPathRoute Set to true if the supplied pattern only matches the path, false if it matches the whole URL.
     * @return true if the request should be handled by the service worker.
     */
    Prefetcher.prototype.shouldIntercept = function (pattern, isPartialPathRoute) {
        var _this = this;
        return function (_a) {
            var url = _a.url;
            var cacheHost = _this.cacheHost;
            var isCorrectOrigin = cacheHost && isPartialPathRoute ? url.origin === cacheHost.origin : true;
            var isMatch = isPartialPathRoute ? pattern.test(url.pathname) : pattern.test(url.toString());
            return isCorrectOrigin && isMatch;
        };
    };
    /**
     * Creates a workbox route.
     * @param pattern The route pattern
     * @param options The caching options specified within the route definition
     * @param origMethod If originally called with options.method !== 'GET', this value
     *  is passed to ensure that a ChangeMethodPlugin is added for the registered GET route
     */
    Prefetcher.prototype.createRoute = function (pattern, options, origMethod) {
        var optMethod = options.method, maxAgeSeconds = options.maxAgeSeconds, isPartialPathRoute = options.isPartialPathRoute;
        var method = (optMethod || 'get').toLowerCase();
        // all routes must have a GET route registered, so Workbox can properly look
        // it up when running workbox-routing/Router#findMatchingRoute:
        if (method !== 'get') {
            this.createRoute(pattern, __assign(__assign({}, options), { method: 'get' }), method);
        }
        var prefetchPlugin = new PrefetchPlugin_1.default({
            incrementHits: this.incrementHits.bind(this),
            incrementMisses: this.incrementMisses.bind(this),
            cacheHost: this.cacheHost,
        });
        var plugins = __spread([
            this.cacheableResponsePlugin,
            new workbox_expiration_1.ExpirationPlugin({
                maxAgeSeconds: maxAgeSeconds,
            }),
            prefetchPlugin
        ], (this.config.plugins || []));
        if (method !== 'get' || origMethod) {
            // ChangeMethodPlugin must come before PrefetchPlugin so body is retained:
            plugins.splice(plugins.indexOf(prefetchPlugin), 0, new ChangeMethodPlugin_1.default([pattern]));
        }
        workbox_routing_1.registerRoute(this.shouldIntercept(pattern, isPartialPathRoute), new workbox_strategies_1.CacheFirst({
            cacheName: constants_1.CACHE_NAME,
            matchOptions: {
                ignoreVary: true,
            },
            plugins: plugins,
        }), 
        // @ts-ignore
        method.toUpperCase() // HTTPMethods map directly from @xdn/core lowercase => @workbox uppercase
        );
    };
    Prefetcher.prototype.incrementHits = function () {
        this.stats.hits++;
    };
    Prefetcher.prototype.incrementMisses = function () {
        this.stats.misses++;
    };
    return Prefetcher;
}());
exports.default = Prefetcher;
